ouroboros-network/src/Ouroboros/Network/PeerSelection/LedgerPeers.hs

```diff
@@ -0,0 +1,188 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Ouroboros.Network.PeerSelection.LedgerPeers (
+    DomainAddress (..),
+    IP.IP (..),
+    LedgerPeersConsensusInterface (..),
+    RelayAddress (..),
+    PoolStake,
+    runLedgerPeers,
+    TraceLedgerPeers (..),
+    pickPeers,
+    ackPoolStake,
+
+    Socket.PortNumber
+    ) where
+
+
+import           Control.Monad.Class.MonadAsync
+import           Control.Monad.Class.MonadSTM
+import           Control.Monad.Class.MonadTime
+import           Control.Monad.Class.MonadTimer
+import           Control.Tracer (Tracer, traceWith)
+import qualified Data.IP as IP
+import           Data.List (foldl', nub)
+import           Data.List.NonEmpty (NonEmpty (..))
+import qualified Data.List.NonEmpty as NonEmpty
+import           Data.Maybe (isNothing, fromJust)
+import qualified Data.Map.Strict as Map
+import           Data.Map.Strict (Map)
+import           Data.Ratio
+import           Data.Word
+import qualified Network.Socket as Socket
+import           System.Random
+
+import           Cardano.Slotting.Slot (SlotNo)
+import           Ouroboros.Network.PeerSelection.RootPeersDNS (DomainAddress (..))
+import           Ouroboros.Network.Subscription.Ip
+import           Ouroboros.Network.Subscription.Dns
+
+import           Text.Printf
+
+newtype LedgerPeersConsensusInterface m = LedgerPeersConsensusInterface {
+      lpGetPeers :: SlotNo -> STM m (Maybe [(PoolStake, NonEmpty RelayAddress)])
+    }
+
+data TraceLedgerPeers =
+      PickedPeer !RelayAddress !AckPoolStake ! PoolStake
+    | PickedPeers !Word16 ![RelayAddress]
+    | FetchingNewLedgerState !Int
+
+
+instance Show TraceLedgerPeers where
+    show (PickedPeer addr ackStake stake) =
+        printf "PickedPeer %s ack stake %s ( %.04f) relative stake %s ( %.04f )"
+            (show addr)
+            (show ackStake) (fromRational ackStake :: Double)
+            (show stake) (fromRational stake :: Double)
+    show (PickedPeers n peers) =
+        printf "PickedPeers %d %s" n (show peers)
+    show (FetchingNewLedgerState cnt) =
+        printf "Fetching new ledgerstate, %d registered pools"
+            cnt
+
+
+data RelayAddress = RelayAddressDomain DomainAddress
+                  | RelayAddressAddr IP.IP Socket.PortNumber
+                  deriving (Show, Eq, Ord)
+
+-- | The relative stake of the stakepool. A value in the [0, 1] range.
+type PoolStake = Rational
+
+-- | The relative stake of the stakepool and all preceeding pools. A value in the range [0, 1].
+type AckPoolStake = Rational
+
+-- | Convert a list of pools with stake to a Map keyed on the ackumulated stake.
+--
+ackPoolStake :: [(PoolStake, NonEmpty RelayAddress)]
+             -> Map Rational (PoolStake, NonEmpty RelayAddress)
+ackPoolStake pl =
+    let pl' = reRelativeStake pl
+        ackList = snd $ foldl' (\(as, ps) (s, rs) -> (as + s, (as + s, (s, rs)):ps)) (0, []) pl' in
```

Comment:
^ the problem with `foldl'` and deep tuples is that tuples are lazy, so strictness of `foldl'` does not help much.

Reply:
Would the insertion of them into a strict Map below force their evaluation?

Reply:
No, that's the same problem.  Haskell evaluates to weak head normal form, which means it will not pass through a lazy constructor.  You need to use `seq` or `!` to force the accumulators, e.g
```
foldl' (\(as, ps) (s, rs) -> 
            let !as' = as + s
                !ps' = (s, rs) : ps
            in (as', (as' , ps'))
        (0, []) pl'
```
And we also increased sharing by using `as'` twice.

Reply:
why do we need to use `as'` twice in the result?

Reply:
We don't need to use `as'` twice, if the folding function written so that it can handle the `[]`.